File: extensions/spark/kyuubi-spark-connector-tpcds/src/main/java/io/trino/tpcds/row/CatalogSalesRow.java
package io.trino.tpcds.row;

    import io.trino.tpcds.type.Pricing;
    import org.apache.kyuubi.spark.connector.tpcds.row.KyuubiTPCDSTableRowWithNulls;

    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_BILL_ADDR_SK;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_BILL_CDEMO_SK;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_BILL_CUSTOMER_SK;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_BILL_HDEMO_SK;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_CALL_CENTER_SK;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_CATALOG_PAGE_SK;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_ORDER_NUMBER;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_COUPON_AMT;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_EXT_DISCOUNT_AMOUNT;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_EXT_LIST_PRICE;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_EXT_SALES_PRICE;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_EXT_SHIP_COST;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_EXT_TAX;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_EXT_WHOLESALE_COST;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_LIST_PRICE;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_NET_PAID;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_NET_PAID_INC_SHIP;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_NET_PAID_INC_SHIP_TAX;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_NET_PAID_INC_TAX;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_NET_PROFIT;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_QUANTITY;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_SALES_PRICE;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_WHOLESALE_COST;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PROMO_SK;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_SHIP_ADDR_SK;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_SHIP_CDEMO_SK;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_SHIP_CUSTOMER_SK;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_SHIP_DATE_SK;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_SHIP_HDEMO_SK;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_SHIP_MODE_SK;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_SOLD_DATE_SK;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_SOLD_ITEM_SK;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_SOLD_TIME_SK;
    import static io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_WAREHOUSE_SK;

public class CatalogSalesRow
    extends KyuubiTPCDSTableRowWithNulls
{
  private final long csSoldDateSk;
  private final long csSoldTimeSk;
  private final long csShipDateSk;
  private final long csBillCustomerSk;
  private final long csBillCdemoSk;
  private final long csBillHdemoSk;
  private final long csBillAddrSk;
  private final long csShipCustomerSk;
  private final long csShipCdemoSk;
  private final long csShipHdemoSk;
  private final long csShipAddrSk;
  private final long csCallCenterSk;
  private final long csCatalogPageSk;
  private final long csShipModeSk;
  private final long csWarehouseSk;
  private final long csSoldItemSk;
  private final long csPromoSk;
  private final long csOrderNumber;
  private final Pricing csPricing;

  public CatalogSalesRow(long csSoldDateSk,
      long csSoldTimeSk,
      long csShipDateSk,
      long csBillCustomerSk,
      long csBillCdemoSk,
      long csBillHdemoSk,
      long csBillAddrSk,
      long csShipCustomerSk,
      long csShipCdemoSk,
      long csShipHdemoSk,
      long csShipAddrSk,
      long csCallCenterSk,
      long csCatalogPageSk,
      long csShipModeSk,
      long csWarehouseSk,
      long csSoldItemSk,
      long csPromoSk,
      long csOrderNumber,
      Pricing csPricing,
      long nullBitMap)
  {
    super(nullBitMap, CS_SOLD_DATE_SK);
    this.csSoldDateSk = csSoldDateSk;
    this.csSoldTimeSk = csSoldTimeSk;
    this.csShipDateSk = csShipDateSk;
    this.csBillCustomerSk = csBillCustomerSk;
    this.csBillCdemoSk = csBillCdemoSk;
    this.csBillHdemoSk = csBillHdemoSk;
    this.csBillAddrSk = csBillAddrSk;
    this.csShipCustomerSk = csShipCustomerSk;
    this.csShipCdemoSk = csShipCdemoSk;
    this.csShipHdemoSk = csShipHdemoSk;
    this.csShipAddrSk = csShipAddrSk;
    this.csCallCenterSk = csCallCenterSk;
    this.csCatalogPageSk = csCatalogPageSk;
    this.csShipModeSk = csShipModeSk;
    this.csWarehouseSk = csWarehouseSk;
    this.csSoldItemSk = csSoldItemSk;
    this.csPromoSk = csPromoSk;
    this.csOrderNumber = csOrderNumber;
    this.csPricing = csPricing;
  }

  public Pricing getCsPricing()
  {
    return csPricing;
  }

  public long getCsSoldItemSk()
  {
    return csSoldItemSk;
  }

  public long getCsCatalogPageSk()
  {
    return csCatalogPageSk;
  }

  public long getCsOrderNumber()
  {
    return csOrderNumber;
  }

  public long getCsBillCustomerSk()
  {
    return csBillCustomerSk;
  }

  public long getCsBillCdemoSk()
  {
    return csBillCdemoSk;
  }

  public long getCsBillHdemoSk()
  {
    return csBillHdemoSk;
  }

  public long getCsBillAddrSk()
  {
    return csBillAddrSk;
  }

  public long getCsCallCenterSk()
  {
    return csCallCenterSk;
  }

  public long getCsShipCustomerSk()
  {
    return csShipCustomerSk;
  }

  public long getCsShipCdemoSk()
  {
    return csShipCdemoSk;
  }

  public long getCsShipAddrSk()
  {
    return csShipAddrSk;
  }

  public long getCsShipDateSk()
  {
    return csShipDateSk;
  }

  public long getCsSoldDateSk() {
    return csSoldDateSk;
  }

  public long getCsSoldTimeSk() {
    return csSoldTimeSk;
  }

  public long getCsShipHdemoSk() {
    return csShipHdemoSk;
  }

  public long getCsShipModeSk() {
    return csShipModeSk;
  }

  public long getCsWarehouseSk() {
    return csWarehouseSk;
  }

  public long getCsPromoSk() {
    return csPromoSk;
  }

  @Override public Object[] values() {
    return new Object[] {
        getOrNullForKey(csSoldDateSk, CS_SOLD_DATE_SK),
        getOrNullForKey(csSoldTimeSk, CS_SOLD_TIME_SK),
        getOrNullForKey(csShipDateSk, CS_SHIP_DATE_SK),
        getOrNullForKey(csBillCustomerSk, CS_BILL_CUSTOMER_SK),
        getOrNullForKey(csBillCdemoSk, CS_BILL_CDEMO_SK),
        getOrNullForKey(csBillHdemoSk, CS_BILL_HDEMO_SK),
        getOrNullForKey(csBillAddrSk, CS_BILL_ADDR_SK),
        getOrNullForKey(csShipCustomerSk, CS_SHIP_CUSTOMER_SK),
        getOrNullForKey(csShipCdemoSk, CS_SHIP_CDEMO_SK),
        getOrNullForKey(csShipHdemoSk, CS_SHIP_HDEMO_SK),
        getOrNullForKey(csShipAddrSk, CS_SHIP_ADDR_SK),
        getOrNullForKey(csCallCenterSk, CS_CALL_CENTER_SK),
        getOrNullForKey(csCatalogPageSk, CS_CATALOG_PAGE_SK),
        getOrNullForKey(csShipModeSk, CS_SHIP_MODE_SK),
        getOrNull(csWarehouseSk, CS_WAREHOUSE_SK),
        getOrNullForKey(csSoldItemSk, CS_SOLD_ITEM_SK),
        getOrNullForKey(csPromoSk, CS_PROMO_SK),
        getOrNull(csOrderNumber, CS_ORDER_NUMBER),
        getOrNull(csPricing.getQuantity(), CS_PRICING_QUANTITY),
        getOrNull(csPricing.getWholesaleCost(), CS_PRICING_WHOLESALE_COST),
        getOrNull(csPricing.getListPrice(), CS_PRICING_LIST_PRICE),
        getOrNull(csPricing.getSalesPrice(), CS_PRICING_SALES_PRICE),
        getOrNull(csPricing.getExtDiscountAmount(), CS_PRICING_EXT_DISCOUNT_AMOUNT),
        getOrNull(csPricing.getExtSalesPrice(), CS_PRICING_EXT_SALES_PRICE),
        getOrNull(csPricing.getExtWholesaleCost(), CS_PRICING_EXT_WHOLESALE_COST),
        getOrNull(csPricing.getExtListPrice(), CS_PRICING_EXT_LIST_PRICE),
        getOrNull(csPricing.getExtTax(), CS_PRICING_EXT_TAX),
        getOrNull(csPricing.getCouponAmount(), CS_PRICING_COUPON_AMT),
        getOrNull(csPricing.getExtShipCost(), CS_PRICING_EXT_SHIP_COST),
        getOrNull(csPricing.getNetPaid(), CS_PRICING_NET_PAID),
        getOrNull(csPricing.getNetPaidIncludingTax(), CS_PRICING_NET_PAID_INC_TAX),
        getOrNull(csPricing.getNetPaidIncludingShipping(), CS_PRICING_NET_PAID_INC_SHIP),
        getOrNull(csPricing.getNetPaidIncludingShippingAndTax(), CS_PRICING_NET_PAID_INC_SHIP_TAX),
        getOrNull(csPricing.getNetProfit(), CS_PRICING_NET_PROFIT)
    };
  }
}

File: extensions/spark/kyuubi-spark-connector-tpcds/src/main/scala/io/trino/tpcds/row/CatalogSalesRow.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.trino.tpcds.row

import java.util.{List => JList}

import io.trino.tpcds.`type`.{Decimal => TPCDSDecimal}
import io.trino.tpcds.`type`.Pricing
import io.trino.tpcds.Options
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_BILL_ADDR_SK
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_BILL_CDEMO_SK
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_BILL_CUSTOMER_SK
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_BILL_HDEMO_SK
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_CALL_CENTER_SK
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_CATALOG_PAGE_SK
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_ORDER_NUMBER
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_COUPON_AMT
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_EXT_DISCOUNT_AMOUNT
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_EXT_LIST_PRICE
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_EXT_SALES_PRICE
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_EXT_SHIP_COST
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_EXT_TAX
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_EXT_WHOLESALE_COST
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_LIST_PRICE
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_NET_PAID
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_NET_PAID_INC_SHIP
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_NET_PAID_INC_SHIP_TAX
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_NET_PAID_INC_TAX
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_NET_PROFIT
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_QUANTITY
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_SALES_PRICE
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PRICING_WHOLESALE_COST
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_PROMO_SK
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_SHIP_ADDR_SK
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_SHIP_CDEMO_SK
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_SHIP_CUSTOMER_SK
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_SHIP_DATE_SK
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_SHIP_HDEMO_SK
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_SHIP_MODE_SK
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_SOLD_DATE_SK
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_SOLD_ITEM_SK
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_SOLD_TIME_SK
import io.trino.tpcds.generator.CatalogSalesGeneratorColumn.CS_WAREHOUSE_SK
import io.trino.tpcds.generator.GeneratorColumn
import org.apache.spark.sql.catalyst.InternalRow
import org.apache.spark.sql.catalyst.expressions.GenericInternalRow
import org.apache.spark.sql.catalyst.util.{DateTimeUtils, RebaseDateTime}
import org.apache.spark.sql.types.Decimal
import org.apache.spark.unsafe.types.UTF8String

class CatalogSalesRow(
    private val csSoldDateSk: Long,
    private val csSoldTimeSk: Long,
    private val csShipDateSk: Long,
    private val csBillCustomerSk: Long,
    private val csBillCdemoSk: Long,
    private val csBillHdemoSk: Long,
    private val csBillAddrSk: Long,
    private val csShipCustomerSk: Long,
    private val csShipCdemoSk: Long,
    private val csShipHdemoSk: Long,
    private val csShipAddrSk: Long,
    private val csCallCenterSk: Long,
    private val csCatalogPageSk: Long,
    private val csShipModeSk: Long,
    private val csWarehouseSk: Long,
    private val csSoldItemSk: Long,
    private val csPromoSk: Long,
    private val csOrderNumber: Long,
    private val csPricing: Pricing,
    nullBitMap: Long) extends TableRowWithNulls(nullBitMap, CS_SOLD_DATE_SK)
  with KyuubiTableRowWithNulls {

  override val nullBitMapInternal = nullBitMap
  override val firstColumnInternal = CS_SOLD_DATE_SK

  def getCsPricing: Pricing = csPricing

  def getCsSoldItemSk: Long = csSoldItemSk

  def getCsCatalogPageSk: Long = csCatalogPageSk

  def getCsOrderNumber: Long = csOrderNumber

  def getCsBillCustomerSk: Long = csBillCustomerSk

  def getCsBillCdemoSk: Long = csBillCdemoSk

  def getCsBillHdemoSk: Long = csBillHdemoSk

  def getCsBillAddrSk: Long = csBillAddrSk

  def getCsCallCenterSk: Long = csCallCenterSk

  def getCsShipCustomerSk: Long = csShipCustomerSk

  def getCsShipCdemoSk: Long = csShipCdemoSk

  def getCsShipAddrSk: Long = csShipAddrSk

  def getCsShipDateSk: Long = csShipDateSk

  def getCsSoldDateSk: Long = csSoldDateSk

  def getCsSoldTimeSk: Long = csSoldTimeSk

  def getCsShipHdemoSk: Long = csShipHdemoSk

  def getCsShipModeSk: Long = csShipModeSk

  def getCsWarehouseSk: Long = csWarehouseSk

  def getCsPromoSk: Long = csPromoSk

  override def getValues: JList[String] = throw new UnsupportedOperationException()

  def internalRow: InternalRow = {
    new GenericInternalRow(
      Array(
        getLongOrNull(csSoldDateSk, CS_SOLD_DATE_SK),
        getLongOrNull(csSoldTimeSk, CS_SOLD_TIME_SK),
        getLongOrNull(csShipDateSk, CS_SHIP_DATE_SK),
        getLongOrNull(csBillCustomerSk, CS_BILL_CUSTOMER_SK),
        getLongOrNull(csBillCdemoSk, CS_BILL_CDEMO_SK),
        getLongOrNull(csBillHdemoSk, CS_BILL_HDEMO_SK),
        getLongOrNull(csBillAddrSk, CS_BILL_ADDR_SK),
        getLongOrNull(csShipCustomerSk, CS_SHIP_CUSTOMER_SK),
        getLongOrNull(csShipCdemoSk, CS_SHIP_CDEMO_SK),
        getLongOrNull(csShipHdemoSk, CS_SHIP_HDEMO_SK),
        getLongOrNull(csShipAddrSk, CS_SHIP_ADDR_SK),
        getLongOrNull(csCallCenterSk, CS_CALL_CENTER_SK),
        getLongOrNull(csCatalogPageSk, CS_CATALOG_PAGE_SK),
        getLongOrNull(csShipModeSk, CS_SHIP_MODE_SK),
        getLongOrNull(csWarehouseSk, CS_WAREHOUSE_SK),
        getLongOrNull(csSoldItemSk, CS_SOLD_ITEM_SK),
        getLongOrNull(csPromoSk, CS_PROMO_SK),
        getLongOrNull(csOrderNumber, CS_ORDER_NUMBER),
        getIntOrNull(csPricing.getQuantity(), CS_PRICING_QUANTITY),
        getDecimalOrNull(csPricing.getWholesaleCost, CS_PRICING_WHOLESALE_COST, 7, 2),
        getDecimalOrNull(csPricing.getListPrice, CS_PRICING_LIST_PRICE, 7, 2),
        getDecimalOrNull(csPricing.getSalesPrice, CS_PRICING_SALES_PRICE, 7, 2),
        getDecimalOrNull(csPricing.getExtDiscountAmount, CS_PRICING_EXT_DISCOUNT_AMOUNT, 7, 2),
        getDecimalOrNull(csPricing.getExtSalesPrice, CS_PRICING_EXT_SALES_PRICE, 7, 2),
        getDecimalOrNull(csPricing.getExtWholesaleCost, CS_PRICING_EXT_WHOLESALE_COST, 7, 2),
        getDecimalOrNull(csPricing.getExtListPrice, CS_PRICING_EXT_LIST_PRICE, 7, 2),
        getDecimalOrNull(csPricing.getExtTax, CS_PRICING_EXT_TAX, 7, 2),
        getDecimalOrNull(csPricing.getCouponAmount, CS_PRICING_COUPON_AMT, 7, 2),
        getDecimalOrNull(csPricing.getExtShipCost, CS_PRICING_EXT_SHIP_COST, 7, 2),
        getDecimalOrNull(csPricing.getNetPaid, CS_PRICING_NET_PAID, 7, 2),
        getDecimalOrNull(csPricing.getNetPaidIncludingTax, CS_PRICING_NET_PAID_INC_TAX, 7, 2),
        getDecimalOrNull(csPricing.getNetPaidIncludingShipping, CS_PRICING_NET_PAID_INC_SHIP, 7, 2),
        getDecimalOrNull(
          csPricing.getNetPaidIncludingShippingAndTax,
          CS_PRICING_NET_PAID_INC_SHIP_TAX,
          7,
          2),
        getDecimalOrNull(csPricing.getNetProfit, CS_PRICING_NET_PROFIT, 7, 2)))
  }
}

trait KyuubiTableRowWithNulls { self: TableRowWithNulls =>

  def isNullInternal(column: GeneratorColumn): Boolean = {
    val kBitMask = 1L << (column.getGlobalColumnNumber - firstColumnInternal.getGlobalColumnNumber)
    (nullBitMapInternal & kBitMask) != 0
  }

  def getLongOrNull(value: Long, column: GeneratorColumn): Any = {
    if (isNullInternal(column) || (value == -1L)) null else value
  }

  def getLongOrNull(value: Int, column: GeneratorColumn): Any = {
    if (isNullInternal(column) || (value == -1)) null else value.toLong
  }

  def getDecimalOrNull(
      value: TPCDSDecimal,
      column: GeneratorColumn,
      precision: Int,
      scale: Int): Decimal = {
    if (isNullInternal(column)) null else Decimal(value.getNumber, precision, scale)
  }

  def getDecimalOrNull(
      value: Int,
      column: GeneratorColumn,
      precision: Int,
      scale: Int): Decimal = {
    if (isNullInternal(column)) null
    else {
      val decimal = Decimal(value)
      decimal.changePrecision(precision, scale)
      decimal
    }
  }

  def getIntOrNull(value: Long, column: GeneratorColumn): Any = {
    if (isNullInternal(column) || (value == -1L)) null else value.toInt
  }

  def getIntOrNull(value: Int, column: GeneratorColumn): Any = {
    if (isNullInternal(column) || (value == -1)) null else value
  }

  def getStringOrNullInternal(value: String, column: GeneratorColumn): UTF8String = {
    if (isNullInternal(column) || value == null || value == Options.DEFAULT_NULL_STRING) {
      null
    } else UTF8String.fromString(value)
  }

  def getStringOrNullInternal(value: Boolean, column: GeneratorColumn): UTF8String = {
    if (isNullInternal(column)) {
      null
    } else {
      if (value) KyuubiTableRowWithNulls.TRUE_STRING else KyuubiTableRowWithNulls.FALSE_STRING
    }
  }

  def getDateOrNullFromJulianDays(value: Long, column: GeneratorColumn): Any = {
    if (isNullInternal(column) || value < 0) {
      null
    } else {
      RebaseDateTime.rebaseJulianToGregorianDays(value.toInt) - DateTimeUtils.JULIAN_DAY_OF_EPOCH
    }
  }

  def nullBitMapInternal: Long
  def firstColumnInternal: GeneratorColumn
  def internalRow: InternalRow
}

object KyuubiTableRowWithNulls {
  private val TRUE_STRING = UTF8String.fromString("Y")
  private val FALSE_STRING = UTF8String.fromString("N")
}
